### DISCUSSION TITLE
Superlocation (get nested path to root location)

### DISCUSSION THREAD

**Discussion by strikaco** [evennia/evennia#3168] [Community Contribs & Snippets]
A `key` is in `a chest` in `Room1`. 

Now tell me: where can I find the key?

Other engines (TADS?) have multiple `location` attributes to resolve an object's location in the world at varying degrees of specificity. So does your national post office (`country > ZIP/postal code > city > street address`).

Evennia will only tell you that the key is in `a chest`. That's not very helpful in most cases, since it's the most specific possible answer and to act on this information you still need to know where the `chest` is. If the `chest` is inside a `wardrobe` object, you have to ask the same question to find where the `wardrobe` is. If the `wardrobe` is inside something else...

In some systems, there are maybe two attributes-- one showing the nearest parent object the target object is in (`chest`), and another showing the nearest parent room (`Room1`). Maybe that's the optimal solution, but for now I have an alternate solution more like a mailing address.

The `superlocation` is the path from the nearest parent object or room, to the furthest parent room.

Thus, if you need to know where something is, you generate the superlocation (its "mailing address" from most to least specific location), then walk it backwards to find exactly where it is within the target location (`country > postal code > city > street address > box number` becomes a `room > container > object` path).

```
from evennia.typeclasses.attributes import NAttributeProperty

# utility function
def superlocate(obj, *args, **kwargs) -> tuple:
    """
    Returns full path to root location from a nested object.

    i.e. object1 is held by char1, who is in room1:

        (char1, room1)

    """
    def iter_location_path(obj, *args, **kwargs) -> iter:
        while True:
            if not obj: 
                break
            elif obj.location:
                obj = obj.location
                yield obj
                continue
            break

    return tuple(
        iter_location_path(obj, *args, **kwargs)
    )

# Mixin implementation, automatically updates on move
class Superlocating(object):

    superlocation = NAttributeProperty(default=(), autocreate=True)

    def at_post_move(self, source_location, *args, **kwargs):
        self.ndb.superlocation = superlocate(self)
        if source_location:
            source_location.ndb.superlocation = superlocate(source_location)
        return super().at_post_move(source_location, *args, **kwargs)


from evennia.utils.test_resources import EvenniaTest
class PathTest(EvenniaTest):

    def test_getpath(self):
        # Obj1 on floor in Room1
        path = superlocate(self.obj1)
        self.assertEqual(path, (self.room1,))

        # Move obj1 to char1. Path is char1->room1.
        self.obj1.move_to(self.char1)
        expect = (self.char1, self.room1)
        path = superlocate(self.obj1)
        self.assertEqual(path, expect)
```

**Comment by InspectorCaracal** [comment]
The key is in your pocket. :wink:

I have a couple of similar recursive functions I put together which do slightly different things. The one that's closest to yours is probably my emote volume, where the louder something is, the more recursions it will do and the further "up" it will send the emote. But the one I use the *most* is part of my attached-parts system, since parts can be attached to each other to an arbitrarily nested depth but I need to be able to identify the "top" object that all of the parts are attached to. 

Neither of them actually track the whole path, since I don't personally use that information so it never occurred to me. 🤔

**Reply by strikaco** [reply]
Heh. It never fails...this too was borne from trying to create a broadcast domain for attenuated messaging ;)

**Reply by InspectorCaracal** [reply]
Oh, that's funny! I just put the recursion directly into my emote method instead of patching the object itself, for that one.

**Reply by strikaco** [reply]
That's how it started, and then I found myself duplicating efforts in trying to make both the character and all belongings wet when submerged in water (or burning if set on fire). If the character stowed a wet item in a held container, I'd have to recursively search all container items they were holding too. And if *that* container got stowed...there'd be wet-and-burning items stashed everywhere.

Ideally I need to get this to a point of quickly selecting all wet items held by character at any distance...which is a variant of the query I needed to message nearby characters at a *limited* distance :P

**Reply by InspectorCaracal** [reply]
Ahh. I ran into the same thing when implementing area effects, but I just do a recursive dive into the contents when applying effects "downward" and then let the objects manage their own effects from there. i.e. the wet-or-burning items know that they're wet or burning, so nothing else needs to check for that in order for the effects of being wet or on fire to take place.

**Comment by owllex** [comment]
I also wrote a `get_root_location` function directly on Object to get to the top level location (but not the path).

I suspect enough people have this use case that it might be worth putting directly on Evennia's ObjectParent rather than in a mixin.

**Reply by strikaco** [reply]
That's why I shared this...I'm curious what other use cases there are and how others solved it. My effort to write a proper API was underwhelming ;)

One problem with both of our approaches is lack of queryability. We can get the root location from a given object, but getting all nested objects from a given root location is a *nasty* query at scale.

(I went with full path because I'm trying to write a graph-based location services API.)

**Reply by owllex** [reply]
Yeah, I agree that writing a query for this isn't easy. I feel like the simple approach of just navigating up the location property from the in-game objects will be most efficient since they are almost always guaranteed to be loaded into memory if your game is running and you already have a handle to a leaf item.
